### PR TITLE
h3i: implement close trigger frames

### PR DESCRIPTION
**Pull request by evanrittenhouse**
- [x] Squash commits

Close trigger frames allow the user to specify a list of frames that the h3i client expects to receive over a given connection.

If h3i sees all of the close trigger frames over the course of the connection, it will pre-emptively close the connection with a CONNECTION_CLOSE frame. If h3i does _not_ see all of the trigger frames, the resulting ConnectionSummary will contain a list of the missing target frames for future inspection.

This gives users a way to close tests out without waiting for the idle timeout, or adding Wait/ConnectionClose actions to the end of each test. This should vastly speed up test suites that have a large number of h3i tests.

In the absence of full-fledged integration tests, I modified the `content_length_mismatch` example to expect a `400`, and remove the existing `Wait/CC` actions:
```
-        Action::Wait {
-            wait_type: WaitType::StreamEvent(StreamEvent {
-                stream_id: STREAM_ID,
-                event_type: StreamEventType::Headers,
-            }),
-        },
-        Action::ConnectionClose {
-            error: quiche::ConnectionError {
-                is_app: true,
-                error_code: quiche::h3::WireErrorCode::NoError as u64,
-                reason: vec![],
-            },
-        },
     ];
 
-    let summary =
-        sync_client::connect(config, &actions, None).expect("connection failed");
+    let summary = sync_client::connect(
+        config,
+        &actions,
+        Some(vec![ExpectedFrame::new(
+            0,
+            vec![quiche::h3::Header::new(b":status", b"400")].into(),
+        )]),
+    )
+    .expect("connection failed");
```

Taking a PCAP:
Screenshot 2024-12-19 at 10 27 50 AM

You can see that the client instantly sends the CC frame upon receiving the headers frame, without us needing to specify the action.

### CODE REVIEW

File: h3i/src/client/connection_summary.rs [hunk 1]
```diff
@@ -106,21 +115,105 @@ impl StreamMap {
     /// use quiche::h3::Header;
     /// use std::iter::FromIterator;
     ///
+    /// let mut stream_map = StreamMap::default();
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:118]: huh?

Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs:118]: Ah missed that in the rebase, thanks. I cleaned up another occurrence as well

File: h3i/src/client/connection_summary.rs [hunk 2]
```diff
@@ -106,21 +115,105 @@ impl StreamMap {
...
+    /// ```
+    /// use h3i::client::connection_summary::StreamMap;
+    /// use h3i::frame::EnrichedHeaders;
+    /// use h3i::frame::H3iFrame;
+    /// use quiche::h3::frame::Frame;
+    /// use quiche::h3::Header;
+    /// use std::iter::FromIterator;
+    ///
+    /// let h = Header::new(b"hello", b"world");
+    /// let correct_headers = H3iFrame::Headers(EnrichedHeaders::from(vec![h]));
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:143]: what does correct mean here?

File: h3i/src/client/connection_summary.rs [hunk 3]
```diff
@@ -106,21 +115,105 @@ impl StreamMap {
...
+        F: Fn(&H3iFrame) -> bool,
+    {
+        self.stream(stream_id)
+            .into_iter()
+            .filter(|frame| f(frame))
+            .collect()
+    }
+
+    /// Collect all frames that match the predicate. The ordering is
+    /// non-deterministic.
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:174]: The non-deterministic bit seems annoying. If we swap the underlying collection for a BTreeMap, they would at least be consistently in increasing stream ID order.   WDYT?

File: h3i/src/client/connection_summary.rs [hunk 4]
```diff
@@ -106,21 +115,105 @@ impl StreamMap {
...
             .flatten()
+            .filter(|frame| f(frame))
             .map(Clone::clone)
-            .collect::<Vec<H3iFrame>>()
+            .collect()
     }
 
-    /// Get all frames on a given `stream_id`.
+    /// Flatten all received frames into a single vector. The ordering is
+    /// non-deterministic.
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:216]: similar comment to above

File: h3i/src/client/connection_summary.rs [hunk 5]
```diff
@@ -213,41 +300,110 @@ impl StreamMap {
...
+    pub(crate) fn new(expected_frames: Option<ExpectedFrames>) -> Self {
+        Self {
+            expected_frames,
+            ..Default::default()
+        }
     }
 
     pub(crate) fn insert(&mut self, stream_id: u64, frame: H3iFrame) {
-        self.0.entry(stream_id).or_default().push(frame);
+        if let Some(expected) = self.expected_frames.as_mut() {
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:314]: I'm not following why this inserts a frame into the expected_frames struct by default

File: h3i/src/client/connection_summary.rs [hunk 6]
```diff
@@ -213,41 +300,110 @@ impl StreamMap {
...
+            expected.receive_frame(stream_id, &frame);
+        }
+
+        self.map.entry(stream_id).or_default().push(frame);
+    }
+
+    pub(crate) fn check_expected_frames(&self) -> bool {
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:321]: Lets document this, since its not immediately obvious what it does. Consdier even returning a enum instead of Bool, since that's harder to misinterpret

File: h3i/src/client/connection_summary.rs (hunk 5 above)
Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs:314]: The idea is that if the stream map sees a frame, and was configured with expected frames, we'll always want to update the expected frame state at the same time that we update the stream map state. If no expected frames were configured, this specific block will be skipped

File: h3i/src/client/connection_summary.rs (hunk 3 above)
Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs:174]: That works. It's also worth noting (and my fault for not being clear in the docs) that the per-stream frame ordering is deterministic - the order in which we iterate over the streams themselves is non-deterministic

File: h3i/src/client/connection_summary.rs (hunk 6 above)
Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs:321]: I renamed it to `saw_expected_frames`, similar to the `ExpectedFrames` naming

File: h3i/src/client/connection_summary.rs (hunk 5 above)
Comment (LPardue) [reply h3i/src/client/connection_summary.rs:314]: Yeah sorry I misread what receive_frame does, so consider documenting that function but nothing here needs change

File: h3i/src/frame.rs [hunk 7]
```diff
@@ -485,25 +555,118 @@ impl ExpectedFrame {
...
+        assert!(this.is_equivalent(&other));
+
+        this = ExpectedFrame::new(0, other);
+        other =
+            vec![Header::new(b"hello", b"world"), Header::new(b"go", b"jets")]
+                .into();
+        // `other` does not contain the `go: devils` header, so it's not
+        // equivalent to `this.
+        assert!(!this.is_equivalent(&other));
+    }
```
Comment (LPardue) [inline h3i/src/frame.rs:612]: I'm finding it hard to follow this example. Can we just split into 2 tests or does the sequencing matter?

Comment (evanrittenhouse) [reply h3i/src/frame.rs:621]: Split it into 2 tests

File: h3i/src/client/connection_summary.rs [hunk 8]
```diff
@@ -246,8 +251,116 @@ impl StreamMap {
...
+            is_app,
+            error_code,
+            reason,
+        }) = self.expected_frames.as_ref().map(|ef| &ef.close_with)
+        else {
+            return;
+        };
+
+        let _ = qconn.close(*is_app, *error_code, reason);
+    }
```
Comment (LPardue) [inline h3i/src/client/connection_summary.rs:300]: Can this just be rewritten like

```
if let Some(ef) = &self.expected_frames {
  qconn.close(*ef.is_app, *ef.error_code, reason);
}
```

?

File: h3i/src/client/sync_client.rs [hunk 9]
```diff
@@ -57,6 +58,15 @@ struct SyncClient {
     stream_parsers: StreamParserMap,
 }
 
+impl SyncClient {
+    fn new(expected_frames: Option<ExpectedFrames>) -> Self {
```
Comment (LPardue) [inline h3i/src/client/sync_client.rs:62]: Should this be part of the `Client` trait or no?

File: h3i/src/lib.rs [hunk 10]
```diff
@@ -112,7 +112,7 @@
 //!    ];
 //!
 //!    let summary =
-//!        sync_client::connect(config, &actions).expect("connection failed");
+//!        sync_client::connect(config, &actions, None).expect("connection failed");
```
Comment (LPardue) [inline h3i/src/lib.rs:115]: Some thing here with `None` not being very helpful to readers.

File: h3i/src/main.rs [hunk 11]
```diff
@@ -298,7 +298,8 @@ fn config_from_clap() -> std::result::Result<Config, String> {
 fn sync_client(
     config: Config, actions: &[Action],
 ) -> Result<ConnectionSummary, ClientError> {
-    h3i::client::sync_client::connect(config.library_config, actions)
+    // TODO: CLI doesn't support passing expected frames at the moment
```
Comment (LPardue) [inline h3i/src/main.rs:301]: Guessing the qlog input format doesn't either?

**Review by LPardue — COMMENTED** [review]
Mostly just nits

File: h3i/src/client/sync_client.rs (hunk 9 above)
Comment (evanrittenhouse) [reply h3i/src/client/sync_client.rs:62]: Unfortunately not, the async client also implements `Client` but its `new` function returns an `AsyncClient`

Comment (evanrittenhouse) [reply h3i/src/client/sync_client.rs:62]: Well, we could probably make `Client<T>` but that will be messier without real benefit IMO

Comment (LPardue) [reply h3i/src/client/sync_client.rs:62]: No biggie

File: h3i/src/client/connection_summary.rs [hunk 12]
```diff
@@ -74,22 +75,32 @@ impl Serialize for ConnectionSummary {
...
         state.end()
     }
 }
 
 /// A read-only aggregation of frames received over a connection, mapped to the
 /// stream ID over which they were received.
 #[derive(Clone, Debug, Default, Serialize)]
-pub struct StreamMap(HashMap<u64, Vec<H3iFrame>>);
+pub struct StreamMap {
+    map: HashMap<u64, Vec<H3iFrame>>,
```
[the remark this answers is not in the document]
Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs]: Changed to `stream_frame_map`

File: h3i/src/client/connection_summary.rs [hunk 13]
```diff
@@ -74,22 +75,32 @@ impl Serialize for ConnectionSummary {
...
+        state.serialize_field(
+            "missed_expected_frames",
+            &self.stream_map.missing_expected_frames(),
+        )?;
         state.end()
     }
 }
 
 /// A read-only aggregation of frames received over a connection, mapped to the
 /// stream ID over which they were received.
```
[the remark this answers is not in the document]
Comment (evanrittenhouse) [reply h3i/src/client/connection_summary.rs:87]: Added a note about them being there so that we can process them as the new `H3iFrame`s come in

**Review by LPardue — APPROVED** [review]
LGTM
